File: richos/mobile/native-android/app/src/main/kotlin/dev/richos/android/platform/ReadReplies.kt
package dev.richos.android.platform

import androidx.activity.ComponentActivity
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import dev.richos.android.core.AppState
import dev.richos.android.core.ReadingAnchor
import dev.richos.android.core.protocol.Row
import dev.richos.android.ui.model.ScreenModel
import dev.richos.android.ui.model.UpdateNotice
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.flow.filterNotNull
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/**
 * D04 (native acceptance r1): a reply the person has read leaves the shade, the way Telegram (the
 * CEO's standard) clears a chat's notifications once it is read. Opened from the notification
 * itself, Android already removes that one (auto-cancel); opened any other way (the launcher,
 * Recents, a link, a return from another app) nothing did, so read replies piled up there.
 *
 * "Read" is what the screen shows, decided from core's state alone (so a test reaches it without a
 * device): the conversation is resumed and drawn, and the reply is at or above where the person is
 * reading. That is the newest while the list follows it, else the row the reader stopped at (core's
 * [ReadingAnchor], D02: the newest row showing at the bottom of the screen). A reply below that
 * position, in another conversation, or not yet delivered to this phone keeps its notification:
 * nothing is withdrawn for a reply the person has not seen.
 *
 * Work only while the conversation is on screen, and only when something changes: no timer, no
 * polling, nothing at all while the app is hidden (CEO ruling §81).
 */
object ReadReplies {
    /** The conversation on screen and its replies at or above the reading position, oldest first. */
    data class Read(val threadId: String, val replyIds: List<String>)

    /** What the person has read right now, or null when no conversation is drawn or it has no reply. Pure. */
    fun of(state: AppState): Read? {
        val threadId = state.selectedThreadId ?: return null
        if (!drawn(state)) return null
        val rows = state.messages
        if (rows.isEmpty()) return null
        val upTo = readingIndex(state.readingAnchor, rows)
        val replies = rows.subList(0, upTo + 1).filter { it.role == "rich" }.map { it.id }
        return if (replies.isEmpty()) null else Read(threadId, replies)
    }

    /**
     * The conversation is what the screen draws (RichApp's own branches): paired, not removed from
     * the Mac, no required update, and no sheet or dialog over it (Settings, Forget, an update
     * dialog), which would leave the thread behind a scrim.
     */
    fun drawn(state: AppState): Boolean {
        val model = ScreenModel(app = state)
        return model.pairingStep == null && !model.removedFromMac && state.sheet == null &&
            model.update !is UpdateNotice.Required && model.update !is UpdateNotice.Dialog
    }

    /**
     * The newest of the Mac's rows the person has reached. No anchor: the list follows the newest.
     * An anchor on a photo's further bubble (`<row>#<n>`) is that row. An anchor that is none of the
     * Mac's rows is either one of the person's own messages, drawn below all of them, or a row no
     * longer loaded, where the conversation opens at the newest (RichApp's `savedIndex` of -1).
     */
    internal fun readingIndex(anchor: ReadingAnchor?, rows: List<Row>): Int {
        val id = anchor?.messageId ?: return rows.lastIndex
        val at = rows.indexOfLast { it.id == id || id.startsWith(it.id + "#") }
        return if (at >= 0) at else rows.lastIndex
    }

    /**
     * While [activity] is resumed, each change in what is read withdraws those replies' notifications.
     * Every entrance is a resume: the launcher, Recents, a link, a tapped notification (whose own
     * reply Android removes; the others read with it follow here). Stops at the pause.
     */
    fun follow(activity: ComponentActivity, states: StateFlow<AppState?>) {
        val context = activity.applicationContext
        activity.lifecycleScope.launch {
            activity.repeatOnLifecycle(Lifecycle.State.RESUMED) {
                states.map { it?.let(::of) }.distinctUntilChanged().filterNotNull().collect { read ->
                    // Android's notification service is a binder call; never on the frame's thread.
                    withContext(Dispatchers.IO) { Replies.withdrawRead(context, read) }
                }
            }
        }
    }
}

File: richos/mobile/native-ios/Core/Sources/RichOSCore/Conversation/ReadReplies.swift
import Foundation

/// D04 (native acceptance r1): a reply the person has read leaves Notification Center, the way
/// Telegram (the CEO's standard) clears a chat's notifications once it is read. A tap on a
/// notification removes that one notification; opened any other way (the Home Screen icon, the app
/// switcher, a link, a return from another app) nothing removed any, so read replies piled up.
/// Android's fix is `279bf32d` (`ReadReplies.kt`); this is the same rule, over the iOS core's state.
///
/// "Read" is what the screen shows, decided from the core's state alone, so a test reaches it with
/// no phone: the conversation is drawn (paired, consent given, nothing covering it), and the reply is
/// at or above where the person is reading. That is every loaded reply while the conversation follows
/// the newest. Otherwise it is the replies at or above the saved reading position, which on iPhone is
/// the TOPMOST row showing (`TranscriptView.visibleAnchor`), so everything counted has been scrolled
/// into view. A reply below that, in another conversation, or not yet on this phone is not read and
/// keeps its notification. With no known position nothing is assumed seen.
///
/// The app evaluates this only while it is on screen and acts only when the answer changes (CEO
/// ruling §81): no timer, no polling, nothing while the app is in the background.
public enum ReadReplies {
    /// The conversation on screen, and its replies at or above the reading position, oldest first.
    public struct Read: Equatable, Sendable {
        public var threadID: String
        public var replyIDs: [String]
        public init(threadID: String, replyIDs: [String]) { self.threadID = threadID; self.replyIDs = replyIDs }
    }

    /// What the person has read right now, or `nil` when no conversation is drawn or none of its
    /// replies is in view. Pure.
    public static func of(_ s: AppState) -> Read? {
        guard let threadID = s.mac?.threadID, drawn(s), let upTo = readingIndex(s) else { return nil }
        let replies = s.messages[...upTo].filter { $0.author == .rich }.map(\.id)
        return replies.isEmpty ? nil : Read(threadID: threadID, replyIDs: replies)
    }

    /// The conversation is what the screen draws and nothing covers it: no takeover (pairing, consent,
    /// removed from the Mac, a required update), no sheet or dialog (Settings, Forget, the system's
    /// microphone question, an update dialog, pair-blocked), no scanner.
    static func drawn(_ s: AppState) -> Bool {
        guard s.screen == .conversation, s.sheet == nil, s.scanner == nil else { return false }
        if case .blockedByUnsentWork? = s.pairingProblem { return false }
        return s.update?.prominence != .dialog
    }

    /// The index of the newest message the person has reached: the newest while following; else the
    /// saved reading position's row; `nil` when that position is not a loaded message.
    static func readingIndex(_ s: AppState) -> Int? {
        guard !s.messages.isEmpty else { return nil }
        if s.following { return s.messages.count - 1 }
        guard let id = s.readingAnchor?.messageID else { return nil }
        return s.messages.lastIndex { $0.id == id }
    }
}
